lib/dhcp/client/client.go

```diff
@@ -244,46 +266,128 @@ func (c *client) renew(id ID, ack dhcp4.Packet, cl *dhcp4client.Client) (dhcp4.P
 	return newack, nil
 }
 
-func (c *client) Renew(id ID, ack *dhcp.Packet) (*dhcp.Packet, error) {
-	c.Lock()
-	defer c.Unlock()
+func (c *client) Renew() error {
+	defer trace.End(trace.Begin(""))
 
-	log.Debugf("renewing IP %s", ack.YourIP())
+	if c.ack == nil {
+		return fmt.Errorf("no ack packet, call Request first")
+	}
 
-	cl, err := c.newClient(ack)
+	cl, err := c.newClient()
 	if err != nil {
-		return nil, err
+		return err
 	}
 	defer cl.Close()
 
-	p := dhcp4.Packet(ack.Packet)
 	var newack dhcp4.Packet
 	err = withRetry(func() error {
 		var err error
-		newack, err = c.renew(id, p, cl)
+		newack, err = c.renew(cl)
 		return err
 	})
 
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return dhcp.NewPacket([]byte(newack)), nil
+	c.ack = newack
+	return nil
 }
 
-func (c *client) Release(ack *dhcp.Packet) error {
-	c.Lock()
-	defer c.Unlock()
+func (c *client) Release() error {
+	defer trace.End(trace.Begin(""))
 
-	log.Debugf("releasing IP %s", ack.YourIP())
+	if len(c.ack) == 0 {
+		return fmt.Errorf("no ack packet, call Request first")
+	}
 
-	cl, err := c.newClient(ack)
+	cl, err := c.newClient()
 	if err != nil {
 		return err
 	}
 	defer cl.Close()
 
 	return withRetry(func() error {
-		return cl.Release(dhcp4.Packet(ack.Packet))
+		return cl.Release(c.ack)
 	})
 }
+
+// SetParamterRequestList sets the DHCP parameter request list
+// per RFC 2132, section 9.8
+func (c *client) SetParameterRequestList(params ...byte) {
+	defer trace.End(trace.Begin(""))
+
+	c.params = make([]byte, len(params))
+	copy(c.params, params)
+	log.Debugf("c.params=%#v", c.params)
+}
+
+// setOptions sets dhcp options on a dhcp packet
+func (c *client) setOptions(p *dhcp4.Packet) (*dhcp4.Packet, error) {
+	defer trace.End(trace.Begin(""))
+
+	dirty := false
+	opts := (*p).ParseOptions()
+
+	// the current parameter request list
+	rl := opts[dhcp4.OptionParameterRequestList]
+	params := make([]byte, len(c.params))
+	copy(params, c.params)
+	if len(rl) > 0 {
+		// figure out if there are any new parameters
+		for i := 0; i < len(params); {
+			if bytes.IndexByte(rl, params[i]) != -1 {
```

Comment:
I think it can be done a little better:
```
output_params := make([]byte, len(c.params), 0)
for _, p := range params {
    if bytes.IndexByte(rl, p) < 0 {
        output_params = append(output_params, p)
    }
}
```